File: ReactNative/appreveal/ios/CapturedRequest.swift
// Models for captured network traffic

import Foundation

public struct CapturedRequest: Codable {
    public let id: String
    public let method: String
    public let url: String
    public let statusCode: Int?
    public let startTime: Date
    public let endTime: Date?
    public let duration: TimeInterval?
    public let requestHeaders: [String: String]
    public let responseHeaders: [String: String]?
    public let requestBodySize: Int?
    public let responseBodySize: Int?
    public let error: String?
    public let redirectCount: Int

    public init(
        id: String = UUID().uuidString,
        method: String,
        url: String,
        statusCode: Int? = nil,
        startTime: Date = Date(),
        endTime: Date? = nil,
        duration: TimeInterval? = nil,
        requestHeaders: [String: String] = [:],
        responseHeaders: [String: String]? = nil,
        requestBodySize: Int? = nil,
        responseBodySize: Int? = nil,
        error: String? = nil,
        redirectCount: Int = 0
    ) {
        self.id = id
        self.method = method
        self.url = url
        self.statusCode = statusCode
        self.startTime = startTime
        self.endTime = endTime
        self.duration = duration
        self.requestHeaders = CapturedRequest.redactSensitiveHeaders(requestHeaders)
        self.responseHeaders = responseHeaders
        self.requestBodySize = requestBodySize
        self.responseBodySize = responseBodySize
        self.error = error
        self.redirectCount = redirectCount
    }

    // Redact sensitive headers
    private static let sensitiveHeaders: Set<String> = [
        "authorization", "cookie", "set-cookie",
        "x-api-key", "x-auth-token", "proxy-authorization"
    ]

    private static func redactSensitiveHeaders(_ headers: [String: String]) -> [String: String] {
        headers.mapValues { value in
            headers.keys.contains(where: { sensitiveHeaders.contains($0.lowercased()) })
                ? "[REDACTED]"
                : value
        }
    }
}

File: Android/appreveal/src/main/kotlin/com/appreveal/network/CapturedRequest.kt
package com.appreveal.network

/**
 * Captured Server-Sent Event frame.
 */
data class CapturedSSEEvent(
    val id: String? = null,
    val event: String? = null,
    val data: String,
    val retry: Long? = null,
    val timestamp: Long = System.currentTimeMillis(),
)

/**
 * Network capture options for automatic client integrations.
 */
data class NetworkCaptureConfig(
    val maxBodyBytes: Long = 256L * 1024L,
    val maxSSEEvents: Int = 200,
    val redactedHeaders: Set<String> = CapturedRequest.defaultSensitiveHeaders,
    val captureBodies: Boolean = true,
)

/**
 * Captured network request data. Matches iOS CapturedRequest shape.
 */
data class CapturedRequest(
    val id: String =
        java.util.UUID
            .randomUUID()
            .toString(),
    val method: String,
    val url: String,
    val statusCode: Int? = null,
    val startTime: Long = System.currentTimeMillis(),
    val endTime: Long? = null,
    val duration: Double? = null,
    val requestHeaders: Map<String, String> = emptyMap(),
    val responseHeaders: Map<String, String>? = null,
    val requestBodySize: Int? = null,
    val responseBodySize: Int? = null,
    val requestBody: String? = null,
    val responseBody: String? = null,
    val requestBodyTruncated: Boolean = false,
    val responseBodyTruncated: Boolean = false,
    val sseEvents: List<CapturedSSEEvent> = emptyList(),
    val isStreaming: Boolean = false,
    val error: String? = null,
    val redirectCount: Int = 0,
) {
    companion object {
        val defaultSensitiveHeaders =
            setOf(
                "authorization",
                "cookie",
                "set-cookie",
                "x-api-key",
                "x-auth-token",
                "proxy-authorization",
            )

        fun redactSensitiveHeaders(
            headers: Map<String, String>,
            redactedHeaders: Set<String> = defaultSensitiveHeaders,
        ): Map<String, String> =
            headers.mapValues { (key, value) ->
                redactHeader(key, value, redactedHeaders)
            }

        fun redactHeader(
            key: String,
            value: String,
            redactedHeaders: Set<String> = defaultSensitiveHeaders,
        ): String = if (redactedHeaders.contains(key.lowercase())) "[REDACTED]" else value
    }

    /**
     * Returns a copy with sensitive headers redacted.
     */
    fun withRedactedHeaders(redactedHeaders: Set<String> = defaultSensitiveHeaders): CapturedRequest =
        copy(
            requestHeaders = redactSensitiveHeaders(requestHeaders, redactedHeaders),
            responseHeaders = responseHeaders?.let { redactSensitiveHeaders(it, redactedHeaders) },
        )
}
